File: backend/src/main/java/com/tmk/vtcmanager/application/domain/cotisation/EncaissementCotisation.java
package com.tmk.vtcmanager.application.domain.cotisation;

import com.tmk.vtcmanager.application.domain.operation.ModePaiement;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class EncaissementCotisation {

    private Long id;
    private Long ligneCotisationId;
    private Long operationFinanciereId;
    private BigDecimal montant;
    private ModePaiement modeEncaissement;
    private LocalDate dateEncaissement;
    private String reference;
    private String commentaire;

    /** Annulé (jamais supprimé) : ignoré des recalculs et des agrégats. */
    private LocalDateTime annuleLe;
    private String annulePar;
    private String motifAnnulation;

    /**
     * Renseigné à la lecture seulement : faux si la date de ce versement ne
     * peut plus être corrigée — ligne annulée ou arrêtée, versement extourné,
     * période close, caisse comptée. Le client s'en sert pour ne pas proposer
     * un geste voué au refus.
     */
    private Boolean dateModifiable;
    /** Ce qui ferme la correction de date, en français. Null quand elle est ouverte. */
    private String motifDateNonModifiable;
}

File: mobile/lib/features/cotisation/domain/entities/encaissement_cotisation.dart
enum ModePaiementCotisation {
  especes,
  mobileMoney;

  static ModePaiementCotisation fromJson(String v) => switch (v) {
        'ESPECES'      => especes,
        'MOBILE_MONEY' => mobileMoney,
        _              => especes,
      };

  String toJson() => switch (this) {
        especes      => 'ESPECES',
        mobileMoney  => 'MOBILE_MONEY',
      };

  String get label => switch (this) {
        especes     => 'Espèces',
        mobileMoney => 'Mobile Money',
      };
}

class EncaissementCotisation {
  final int? id;
  final int ligneCotisationId;
  final int? operationFinanciereId;
  final double montant;
  final ModePaiementCotisation modeEncaissement;
  final DateTime dateEncaissement;
  final String? reference;
  final String? commentaire;

  /// Renseignés si le versement a été extourné : il reste dans la liste — il a
  /// eu lieu — mais ne compte plus dans ce que la ligne a encaissé.
  final DateTime? annuleLe;
  final String? motifAnnulation;

  bool get estAnnule => annuleLe != null;

  const EncaissementCotisation({
    this.id,
    required this.ligneCotisationId,
    this.operationFinanciereId,
    required this.montant,
    required this.modeEncaissement,
    required this.dateEncaissement,
    this.reference,
    this.commentaire,
    this.annuleLe,
    this.motifAnnulation,
  });
}
